Lib/test/test_asyncio/test_queues.py

```diff
@@ -522,5 +522,145 @@ class PriorityQueueJoinTests(_QueueJoinTestMixin, unittest.IsolatedAsyncioTestCa
     q_class = asyncio.PriorityQueue
 
 
+class _QueueShutdownTestMixin:
+    q_class = None
+
+    @staticmethod
+    async def _ping_awaitable(a):
+        try:
+            await asyncio.wait_for(asyncio.shield(a), 0.01)
+        except TimeoutError:
+            pass
+
+    def assertRaisesShutdown(self, msg="Didn't appear to shut-down queue"):
+        return self.assertRaises(asyncio.QueueShutDown, msg=msg)
+
+    async def test_format(self):
+        q = self.q_class()
+        q._is_shutdown = True
+        self.assertEqual(q._format(), 'maxsize=0 shutdown')
+
+    async def test_shutdown_empty(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        join_task = loop.create_task(q.join())
+        q.shutdown()
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            await q.get()
+        await join_task
+
+    async def test_shutdown_nonempty(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        q.put_nowait("data")
+        join_task = loop.create_task(q.join())
+        q.shutdown()
+
+        self.assertEqual(await q.get(), "data")
+
+        await self._ping_awaitable(join_task)
+        self.assertFalse(join_task.done())
+
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            q.put_nowait("data")
+
+        with self.assertRaisesShutdown():
+            await q.get()
+        with self.assertRaisesShutdown():
+            q.get_nowait()
+
+        q.task_done()
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+        await join_task
+
+    async def test_shutdown_immediate(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        q.put_nowait("data")
+        join_task = loop.create_task(q.join())
+        q.shutdown(immediate=True)
+
+        await self._ping_awaitable(join_task)
```

Comment:
In the `self._ping_awaitable` method, which block is executed for each of the 2 calls (lines 592 and 610)?  The `try` block or the `except` block.
Should we check?

Reply:
The point of `_ping_awaitable` is to have the async loop run the code at all: it doesn't care how or what it raises or returns (that's checked elsewhere in the test, eg `assertTrue(join_task.done())` and `await join_task`)

Reply:
Okay, so can I consider this to be equivalent to an `await asyncio.sleep(0)` instruction?
Or if I'm wrong, what's the difference?

Reply:
`await asyncio.sleep(0)` will yield to the the next task scheduled, which after future code edits may not be `join_task`. This makes it explicit that `join_task` is depended on, and forces it to be started (if not already). I'll add this as a comment

Reply:
I see, I didn't know `await asyncio.sleep(0)` yields to all scheduled tasks